### PR TITLE
[aarch64] atan2 intrinsic lowering (p5)

### PR DESCRIPTION
**Pull request by tex3d**
This change is part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

- `VecFuncs.def`: define intrinsic to sleef/armpl mapping
- `LegalizerHelper.cpp`: add missing fewerElementsVector handling for the new atan2 intrinsic
- `AArch64ISelLowering.cpp`: Add arch64 specializations for lowering like neon instructions
- `AArch64LegalizerInfo.cpp`: Legalize atan2.

Part 5 for Implement the atan2 HLSL Function #70096.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir [hunk 1]
```diff
@@ -0,0 +1,255 @@
+# RUN: llc -verify-machineinstrs -mtriple aarch64--- \
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir:1]: Why does this file not start with:
; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
?
Look at other legalize-foo.mir files for prior art in the same directory.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir:1]:
```
> ./llvm/utils/update_mir_test_checks.py llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir
```

File: llvm/test/CodeGen/AArch64/replace-with-veclib-armpl.ll [hunk 2]
```diff
@@ -15,7 +15,7 @@ declare <vscale x 2 x double> @llvm.cos.nxv2f64(<vscale x 2 x double>)
 declare <vscale x 4 x float> @llvm.cos.nxv4f32(<vscale x 4 x float>)
```
Comment (farzonl) [inline llvm/test/CodeGen/AArch64/replace-with-veclib-armpl.ll:15]: I was expecting 
- `llvm/test/CodeGen/AArch64/replace-with-veclib-sleef.ll` 
- `llvm/test/CodeGen/AArch64/replace-with-veclib-sleef-scalable.ll`
- `llvm/test/CodeGen/AArch64/vec-libcalls.ll`
- `llvm/test/Transforms/LoopVectorize/AArch64/veclib-calls-libsystem-darwin.ll`
- `llvm/test/Transforms/LoopVectorize/AArch64/veclib-intrinsic-calls.ll`
I think you may have done some of these in another PR. But want to know what the plan is to do the rest?

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir (hunk 1 above)
Comment (tex3d) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir:2]: Before, this test was manually adapted from the pattern used for some of the other arc/hyperbolic intrinsics (such as legalize-atan.mir).  I updated the test to the result of using update_mir_test_checks.py instead.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir:2]: Thanks. Automatism is great and the header gives us confidence that the check lines are correct.

**Review by farzonl — APPROVED** [review]

**Comment by tex3d** [comment]
Failures in buildkite are unrelated.  I might update the branch just to kick it off again, hopefully with that issue fixed.

**Review by bogner — APPROVED** [review]

**Comment by tschuett** [comment]
Known Windows issue:
https://github.com/llvm/llvm-project/issues/112961

**Comment by tex3d** [comment]
@tschuett 
> Known Windows issue: #112961

Thanks, I rebased, so hopefully those pipelines will pass now.  If the change looks good to you, would you like to approve?

**Pull request merged**